bloom.go

```diff
@@ -0,0 +1,890 @@
+	b.add(b.hasher(data))
+}
+
+func (b *bloom) maybeHas(xs []uint32) bool {
+	for _, x := range xs {
+		if b.bits[int(x/8)%len(b.bits)]&(1<<(x%8)) == 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func (b *bloom) maybeHasBytes(xs []byte) bool {
+	return b.maybeHas(b.hasher(xs))
+}
+
+func (b *bloom) load() float64 {
+	total := 0
+	for _, x := range b.bits {
+		total += bits.OnesCount8(x)
+	}
+	return float64(total) / float64(len(b.bits)*8)
+}
+
+// shrinkToSize returns a resized bloom filter with a bit density close to target.
+// This exploits the fact that a test for a probe x in the bloom filter is actually
+// a test for bit x%len, and a bloom filter of size newlen that divides len is easily
+// derived by oring the bits together len/newlen times. This works because
+// x%newlen == x%len%newlen iff newlen divides len, so we can shrink the bloom filter
+// without having the original probes or keys! This functionality lets us construct
+// a bloom filter while only having an upper bound on cardinality, instead of having
+// to have a separate, expensive input-counting phase.
+func (b *bloom) shrinkToSize(target float64) bloom {
+	if target <= 0.0 || target >= 1.0 {
+		return *b
+	}
+
+	// shrinking sets each output bit to the OR-ed together
+	// output of k=`factor` bits that are set with probability
+	// x=`b.load()`. We want to achieve a target load `y`.
+	//
+	// The probability that a bit is set is one minus the probability
+	// that its inputs are all unset-- 1-(1-x)^k. To get k given y,
+	// https://www.wolframalpha.com/input/?i=solve+for+k+in+y%3D1-%281-x%29%5Ek
+	// => k=log(1-y)/log(1-x)
+	factor := int(math.Log(1-target) / math.Log(1-b.load()))
+
+	// We can only shrink the bloom filter to a size
+	// that is a factor of the input size. This is made easier by bloomSizeBase
+	// being highly composite.
+	for factor > 0 && len(b.bits)%factor != 0 {
+		factor--
+	}
+
+	if factor <= 1 {
+		return *b
+	}
+	out := bloom{b.hasher, make([]uint8, len(b.bits)/factor)}
+	j := 0
+	for i := 0; i < len(b.bits); i++ {
+		out.bits[j] |= b.bits[i]
+		j++
+		if j >= len(out.bits) {
+			j = 0
+		}
+	}
+
+	return out
+}
+
+func (b *bloom) GobEncode() ([]byte, error) {
+	out := make([]byte, len(b.bits)+2)
+	out[0] = 1
+	out[1] = bloomHasherIds[reflect.ValueOf(b.hasher).Pointer()]
+	copy(out[2:], b.bits)
+	return out, nil
+}
+
+func (b *bloom) GobDecode(buf []byte) error {
+	if len(buf) < 2 || buf[0] != 1 {
+		return errors.New("invalid bloom filter encoding (wrong size/version)")
+	}
+	if buf[1] <= 0 || int(buf[1]) > len(bloomHashers) {
+		return errors.New("invalid bloom filter encoding (unknown hasher type)")
+	}
+	b.hasher = bloomHashers[buf[1]-1]
+	b.bits = make([]byte, len(buf)-2)
+	copy(b.bits, buf[2:])
+	return nil
+}
+
+var bloomHasherIds = map[uintptr]byte{
+	reflect.ValueOf(bloomHasherCRCBlocked64B8K3).Pointer(): 1,
+}
+
+var bloomHashers = []bloomHash{
+	bloomHasherCRCBlocked64B8K3,
+}
+
+func bloomHasherCRC(in []byte) []uint32 {
```

Comment:
TODO @stefanhengl and I have reviewed everything except the file from this point and the tests.